mopidy/internal/path.py

```diff
@@ -64,199 +60,54 @@ def path_to_uri(path):
 
     Returns a file:// URI as an unicode string.
     """
-    if isinstance(path, compat.text_type):
-        path = path.encode('utf-8')
-    if compat.PY2:
-        path = urllib.parse.quote(path)
-    else:
-        path = urllib.parse.quote_from_bytes(path)
-    return urllib.parse.urlunsplit(('file', '', path, '', ''))
+    return pathlib.Path(path).as_uri()
 
 
 def uri_to_path(uri):
     """
     Convert an URI to a OS specific path.
-
-    Returns a bytestring, since the file path can contain chars with other
-    encoding than UTF-8.
-
-    If we had returned these paths as unicode strings, you wouldn't be able to
-    look up the matching dir or file on your file system because the exact path
-    would be lost by ignoring its encoding.
     """
     if compat.PY2:
         if isinstance(uri, compat.text_type):
             uri = uri.encode('utf-8')
-        return urllib.parse.unquote(urllib.parse.urlsplit(uri).path)
+        bytes_path = urllib.parse.unquote(urllib.parse.urlsplit(uri).path)
+        return pathlib.Path(bytes_path)
     else:
-        return urllib.parse.unquote_to_bytes(urllib.parse.urlsplit(uri).path)
-
-
-def split_path(path):
-    if not isinstance(path, bytes):
-        raise TypeError('path is not a bytestring: %r' % path)
-    parts = []
-    while True:
-        path, part = os.path.split(path)
-        if part:
-            parts.insert(0, part)
-        if not path or path == b'/':
-            break
-    return parts
+        bytes_path = urllib.parse.unquote_to_bytes(
+            urllib.parse.urlsplit(uri).path)
+        unicode_path = bytes_path.decode('utf-8', 'surrogateescape')
+        return pathlib.Path(unicode_path)
 
 
 def expand_path(path):
-    # TODO: document as we want people to use this.
-    if not isinstance(path, bytes):
-        raise TypeError('path is not a bytestring: %r' % path)
+    path = str(pathlib.Path(path))
     for xdg_var, xdg_dir in XDG_DIRS.items():
-        var = ('$' + xdg_var).encode('utf-8')
-        path = path.replace(var, xdg_dir)
-    if b'$' in path:
+        path = path.replace('$' + xdg_var, str(xdg_dir))
+    if '$' in path:
```

Comment:
If `path` is bytes in PY2 and `'$'` is Unicode literal, won't this `UnicodeDecodeError`?

Reply:
Python 2 allows way too much:

``` python
In [20]: path = b'/foo/$XDG_VAR/bar'

In [21]: path.replace(u'$XDG_VAR', str('baz'))
Out[21]: u'/foo/baz/bar'
```

On Python 3 all the strings here should be unicode text objects.

Reply:
But can't `xdg_dir` contain non-ASICC?
```
>>> path.replace(u'$XDG_VAR', str('baz'))
u'/foo/baz/bar'
>>> path.replace(u'$XDG_VAR', str('båz'))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 1: ordinal not in range(128)
```

Reply:
Yes it can, for example if a user has a non-ASCII character in their homedir path.